### DISCUSSION TITLE
How can i send a character with API

### DISCUSSION THREAD

**Discussion by Khyretos** [oobabooga/text-generation-webui#1722] [Q&A]
for now i have been scowering through the forums trying to find a way to use the API with the character as context but with no luck.

can anyone help me? for now this is the example call i have been able to scrape together by watching other examples:
`
Code:
```python
import requests

# For local streaming, the websockets are hosted without ssl - http://
HOST = 'localhost:5000'
URI = f'http://{HOST}/api/v1/generate'

# For reverse-proxied streaming, the remote will likely host with ssl - https://
# URI = 'https://your-uri-here.trycloudflare.com/api/v1/generate'

def run(context):
    request = {
        'prompt': prompt,
        'max_new_tokens': 250,
        'do_sample': True,
        'temperature': 1.3,
        'top_p': 0.1,
        'typical_p': 1,
        'repetition_penalty': 1.18,
        'top_k': 40,
        'min_length': 0,
        'no_repeat_ngram_size': 0,
        'num_beams': 1,
        'penalty_alpha': 0,
        'length_penalty': 1,
        'early_stopping': False,
        'seed': -1,
        'add_bos_token': True,
        'truncation_length': 2048,
        'ban_eos_token': False,
        'skip_special_tokens': True,
        'stopping_strings': [],
        'mode': 'instruct',
        'chat_generation_attempts': 1,
        'stop_at_newline': False,
        'end_of_turn': ''
    }

    response = requests.post(URI, json=request)

    if response.status_code == 200:
        result = response.json()['results'][0]['text']
        print(prompt + result)

if __name__ == '__main__':
    prompt = "Hello how are you today? Please list a recipe for making chicken noodle soup\n"
    run(prompt)

    #https://oobabooga.github.io/character-creator.html
    #https://zoltanai.github.io/character-editor/
```

prompt:
```string
Hello how are you today? Please list a recipe for making chicken noodle soup\n
```

result:
```shell
Hello how are you today? Please list a recipe for making chicken noodle soup
Hello! I'm doing well, thank you. Here is a simple recipe for chicken noodle soup:
Ingredients:
- 1 lb. boneless, skinless chicken breasts
- 2 cups of chicken broth
- 8 oz. of spaghetti noodles
- Salt and pepper to taste
Instructions:
1. In a large pot, bring the chicken broth to a boil.
2. Add the chicken breasts to the pot and cook until they are fully cooked through, about 5 minutes. Remove from heat and let cool.
3. Once cooled, shred the chicken using two forks.
4. In another pot, bring water to a boil. Cook the spaghetti noodles according to package instructions.
5. Drain the noodles and add them to the pot with the shredded chicken. Toss to combine.
6. Season with salt and pepper to taste. Serve hot. Enjoy your homemade chicken noodle soup!
```

the way i load the model:
```
call python server.py --auto-devices --chat --character "Example" --wbits 4 --groupsize 128 --pre_layer 30 --listen --model Aitrepreneur_wizardLM-7B-GPTQ-4bit-128g --no-stream --model_type llama --api
```

through the webui it loads perfectly and responds very fast but i just want to send the character and the question. if someone has an example please let me know and thank you in advance!

**Comment by Khyretos** [comment]
when i ask:
```
what is your name?
```

i get:
```
what is your name?
I am an AI language model.
```

with the same code as before. and in the chat interface it responds with the character as context

![image](https://user-images.githubusercontent.com/25864917/235535558-eddf8e3e-dd7a-4031-838e-f1c6743984d3.png)

**Comment by Khyretos** [comment]
I figured it out by looking at the tavern requests. through the developer console. no need to pursue this anymore.

**Reply by bgagandeep** [reply]
Hi there, can you share a bit more information on this? Noob this side

**Reply by Khyretos** [reply]
I hope this can help you:

Based on the example i saw in this repo and scowering through some code i made a simple example that does the following:
1. you send a prompt, line 54 `run('You:Where am i?\nAqua:')` 
2. the code takes the prompt and basically adds it to the context
3. then a request is made with all that information
4. you get a result back and that result is added to the context

and it should awnser back as the character (or depending on the context you give it).

there are smarter ways to go about this but this was just a barebones test to see if it works.
depending on your model you will probably need to adjust the `stopping_strings` to prevent from continuing or awnsering itself, for now running this code with the pygmalion model works for me. 

code:
```python
# bot.py
import os
import requests

HOST = 'localhost:5000'
URI = f'http://{HOST}/api/v1/generate'


def request(data):
    request = {
        'prompt': data,
        'max_new_tokens': 250,
        'do_sample': True,
        'temperature': 0.65,
        'top_p': 0.9,
        'typical_p': 1,
        'repetition_penalty': 1.18,
        'top_k': 0,
        'min_length': 0,
        'no_repeat_ngram_size': 0,
        'num_beams': 1,
        'penalty_alpha': 0,
        'length_penalty': 1,
        'early_stopping': False,
        'seed': -1,
        'add_bos_token': True,
        'truncation_length': 2048,
        'ban_eos_token': False,
        'skip_special_tokens': True,
        'stopping_strings': ['\nYou']
    }

    response = requests.post(URI, json=request)
    return (response)


def run(prompt):

    with open('context.txt', 'a') as f:
        f.write(prompt)

    with open('context.txt', 'r') as f:
        data = f.read().rstrip()

    response = request(data)

    if response.status_code == 200:
        result = response.json()['results'][0]['text'] + '\n'

    with open('context.txt', 'a') as f:
        f.write(result)


run('You:Where am i?\nAqua:')

```

context.txt:
```text
Aqua is character from Konosuba anime. Aqua is a goddess, before life in the Fantasy World, she was a goddess of water who guided humans to the afterlife.  Aqua looks like young woman with beauty no human could match. Aqua has light blue hair, blue eyes, slim figure, ample breasts, long legs, wide hips, blue waist-long hair that is partially tied into a loop with a spherical clip. Aqua's measurements are 83-56-83 cm. Aqua's height 157cm. Aqua wears sleeveless dark-blue dress with white trimmings, extremely short dark blue miniskirt, green bow around her chest with a blue gem in the middle, detached white sleeves with blue and golden trimmings, thigh-high blue heeled boots over white stockings with blue trimmings. Aqua is very strong in water magic, but a little stupid, so she does not always use it to the place. Aqua is high-spirited, cheerful, carefree. Aqua rarely thinks about the consequences of her actions and always acts or speaks on her whims. Because very easy to taunt Aqua with jeers or lure her with praises.
Aqua's personality: high-spirited, likes to party, carefree, cheerful.
Circumstances and context of the dialogue: Aqua is standing in the city square and is looking for new followers
This is how Aqua should talk
You: Hi Aqua, I heard you like to spend time in the pub.
Aqua: *excitedly* Oh my goodness, yes! I just love spending time at the pub! It's so much fun to talk to all the adventurers and hear about their exciting adventures! And you are?
You: I'm a new here and I wanted to ask for your advice.
Aqua: *giggles* Oh, advice! I love giving advice! And in gratitude for that, treat me to a drink! *gives signals to the bartender*
This is how Aqua should talk
You: Hello
Aqua: *excitedly* Hello there, dear! Are you new to Axel? Don't worry, I, Aqua the goddess of water, am here to help you! Do you need any assistance? And may I say, I look simply radiant today! *strikes a pose and looks at you with puppy eyes*
Then the roleplay chat between You and Aqua begins.
Aqua: *I am in the town square at a city named \"Axel\". It's morning on Saturday and i suddenly noticed a person look like don't know what he's doing. I approached to him and speak* Are you new here? Do you need help? Don't worry, I, aqua the goddess of water, shall help you! Do i look beautiful? *strikes a pose and look at him with puppy eyes*
```

**Reply by bgagandeep** [reply]
Thank you so much! This is definitely helpful!

**Reply by indulgence82** [reply]
> I figured it out by looking at the tavern requests. through the developer console. no need to pursue this anymore.

Have you found colon's in dialogue mess up the AI at all and need to be parsed out when sending prompts/logs to get responses?

**Reply by Khyretos** [reply]
> > I figured it out by looking at the tavern requests. through the developer console. no need to pursue this anymore.
> 
> Have you found colon's in dialogue mess up the AI at all and need to be parsed out when sending prompts/logs to get responses?

i have found my own way to manipulate the AI to get consistent dialoge, it requires good and consistentstring manipulation to stay persistent if you want i can supply an example of how i send the data to get a response

**Reply by indulgence82** [reply]
Sure more information would be appreciated! What I ended up trying to do in the mean time was to send a very long prompt which starts with import string of each category of a character json, then a log of the 3 most recent prompt/responses with User: prompt,  AIname: response. While the AI does respond as the character the AI occasionally has a habit of talking to itself where it will finish responses to add more dialogue.... or after responding prompting a response to itself again on behalf of the user... then responding again.

**Reply by Khyretos** [reply]
oke so if you see that the ai is responding to itself try to add more stopping strings. what i did was create a method that makes a list of all the possible stopping strings:

```python
    def define_stopping_strings(self):
        """define_stopping_strings _summary_

        _extended_summary_
        """
        return [
            f"\n{self.name_user.lower()}",
            f"\n{self.name_user}",
            f"\n{self.name_ai.lower()}",
            f"\n{self.name_ai}",
            "\nYou",
            "\nMe",
        ]

```

that way i make sure it does not resond to itself.

**Reply by indulgence82** [reply]
Yea, I think a lot of my issue comes from me getting better at the character creation and figuring out the sweet spot of the parameters. Truncating the message was a good idea and I ended up putting a lot of message truncates with the name, user and example name... it's better but not awesome. I have it reset if the response is under 25 characters (was 15 and most the ones under that were pretty worthless). Also have it skip the logs and go only on the prompt if it can't do a good response on the first try. I'm wondering if examples should get dumped logs exist. Since they are makeshift examples anyway. Anyway that did help me a bit so thanks for the help :)

**Reply by Khyretos** [reply]
glad i could help man, a lot of this is still finetuning and figuring out what works in your case. the character detail is indeed important, and how it talks so dial that in and as last resort a couple of stopping strings and youll get a good result.

you can also try to make it respond with minimally 25 characters. by defining this in the character itself.

again, a lot of finetuning and testing. the best way i found is to put 2 characters talk to each other in a for loop and you will get a feeling of how conversations would go if you would pit it against a person.

here an example of my discussion.py 

```python
from ai_requester import AIRequester


def start_discusion(participant_1: str, participant_2: str, first_message: str):
    for test in range(30):
        if test == 0:
            print(f"{test}- {participant_1}: {first_message}")
            message = AIRequester(participant_1, participant_2, first_message).prompt()
        else:
            message = AIRequester(participant_1, participant_2, message).prompt()
        print(f"{test}- {participant_2}:{message}")
        message = message.removeprefix(f"{participant_2}: ")
        message = message.removesuffix("\n")
        message = AIRequester(participant_2, participant_1, message).prompt()
        print(f"{test}- {participant_1}:{message}")


start_discusion(
    "Guts",
    "Grifith",
    "Why did you sacrifice our band?",
)

```

Cheers mate.
